src/stdlib/Conversions.scillib

```diff
@@ -0,0 +1,221 @@
+(*   version.                                                                       *)
+(*                                                                                  *)
+(*   scilla is distributed in the hope that it will be useful, but WITHOUT ANY      *)
+(*   WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR  *)
+(*   A PARTICULAR PURPOSE.  See the GNU General Public License for more details.    *)
+(*                                                                                  *)
+(*   You should have received a copy of the GNU General Public License along with   *)
+(*   scilla.  If not, see <http://www.gnu.org/licenses/>.                           *)
+(* ******************************************************************************** *)
+scilla_version 0
+
+import IntUtils
+
+library Conversions
+
+type IntegerEncoding =
+  | LittleEndian
+  | BigEndian
+
+(* does not throw exceptions *)
+let osubstr : ByStr -> Uint32 -> Uint32 -> Option ByStr =
+  fun (bs : ByStr) =>
+  fun (pos : Uint32) =>
+  fun (len : Uint32) =>
+    let length = builtin strlen bs in
+    let shorter = uint32_le len length in
+    match shorter with
+    | True =>
+      let delta = builtin sub length len in
+      let safe_pos = uint32_le pos delta in
+      match safe_pos with
+      | True =>
+        let substr = builtin substr bs pos len in
+        Some {ByStr} substr
+      | False => None {ByStr}
+      end
+    | False => None {ByStr}
+    end
+
+(* Extract out a Scilla type from a ByStr starting at pos. Returns next position. *)
+(* Use the type specific helpers below rather than this function. *)
+let extract_scillaval : forall 'A. forall 'B. (ByStr -> Option 'B) -> ('B -> 'A) -> ByStr -> Uint32 -> Uint32 -> Option (Pair 'A Uint32) =
+  tfun 'A =>
+  tfun 'B =>
+  fun (to_bystrx : ByStr -> Option 'B) =>
+  fun (to_uint : 'B -> 'A) =>
+  fun (bs : ByStr) =>
+  fun (pos : Uint32) =>
+  fun (len : Uint32) =>         (* Byte size of the Uint value. *)
+    let subbs_opt =  osubstr bs pos len in
+    match subbs_opt with
+    | Some subbs =>
+      let subbs_x_opt = to_bystrx subbs in
+      match subbs_x_opt with
+      | Some subbs_x =>
+        let ui = to_uint subbs_x in
+        let next_pos = builtin add pos len in
+        let res = Pair {'A Uint32} ui next_pos in
+        Some {(Pair 'A Uint32)} res
+      | None =>
+        None {(Pair 'A Uint32)}
+      end
+    | None =>
+      None {(Pair 'A Uint32)}
+    end
+
+(* Extracts Uint32 in bs from position pos and returns next position. *)
+let extract_uint32 : IntegerEncoding -> ByStr -> Uint32 -> Option (Pair Uint32 Uint32) =
+  fun (endian : IntegerEncoding) =>
+  fun (bs : ByStr) =>
+  fun (pos : Uint32) =>
+    let to_bystrx = fun (a : ByStr) => builtin to_bystr4 a in
+    let to_uint =
+      fun (a : ByStr4) =>
+        let b = match endian with | LittleEndian => builtin strrev a | BigEndian => a end in
```

Comment:
I think we ought to have a helper function to do this little-endian conversion, because we use it lots of times.

Reply:
I considered that, but seemed like an overkill because it works on different types each type (different `X`s for the `ByStrX`) and that means, an instantiation each time. What we have now gets done in a single line. 
If that's still ok, I'll do this, let me know.

Reply:
Ah, I hadn't noticed that the types were different. Never mind.